src/librustc/mir/mod.rs

```diff
@@ -402,9 +402,11 @@ pub enum BorrowKind {
 ///////////////////////////////////////////////////////////////////////////
 // Variables and temps
 
-newtype_index!(Local, "_");
-
-pub const RETURN_POINTER: Local = Local(0);
+newtype_index!(Local,
+    const {
+        DESCRIPTION = "_",
```

Comment:
Nit: I don't think I would include `DESCRIPTION` in the list of constants, since it's not a constant.. But I sort of like having one big list. Maybe we should just remove the `const` keyword and instead write something like:

```rust
newtype_index!(Local {
    DESCRIPTION = "_",
    RETURN_POINTER = 0,
});
```

Alternatively, put the `const` keyword on the actual constants:

```rust
newtype_index!(Local {
    DESCRIPTION = "_",
    const RETURN_POINTER = 0,
});
```

Reply:
And we could even lowercase the non constants ...

```
newtype_index!(Local {
    debug_name = "_",
    const RETURN_POINTER = 0,
});
```

Reply:
If we go with something like
```rust
newtype_index!(Local {
    DESCRIPTION = "_",
    const RETURN_POINTER = 0,
});
```
Can we add the restriction that `MAX` and `DEBUG_NAME` come first? Coming up with a macro that supports an optional, initial keyword for each definition in the list is going to be much more challenging. It may not be worth the extra effort. I can document that limitation at the macro definition.

@spastorino I would still keep them all caps since their values don't change. I think the `const` keyword will signify that it's an externally accessible definition.

Reply:
For clarification, the challenge comes in defining the macro expression to parse the entire list. Something like
```rust
{ $($(const)* $idents:ident = $constants:expr),+ }
```
Fails with `local ambiguity: multiple parsing options: built-in NTs ident ('idents') or 1 other option.`.
I can change it to $($definitions:tt),* but then I'll need to account for macro variable type casting (things like `@as_expr $e:expr` and `@as_ident $i:ident`) to convert the generic tokens into something more parselable. I'm not sure how long that will take to figure out vs just distinguishing between non-consts first followed by consts.

Reply:
@Nashenas88 

> Can we add the restriction that MAX and DEBUG_NAME come first?

Definitely. This is an internal macro, so it doesn't have to be particularly flexible. I'd say make it so that MAX / DEBUG_NAME must come first and in a particular order.

(Incidentally, usually when I have to struct macros like this, I have a kind of "consume loop" where it peels off tokens from the front and then recursively calls itself on the remaining tokens. But I guess that's what you were trying to avoid.)

Reply:
Somehow I missed this comment and in the meantime implemented a version that does work and is less lines. It doesn't enforce MAX and DEBUG_NAME coming first, though. Should I still add that restriction in?